Research/ResearchUI/iOS/View Controllers/RSDVideoViewController.swift

```diff
@@ -0,0 +1,73 @@
+//
+// 1.  Redistributions of source code must retain the above copyright notice, this
+// list of conditions and the following disclaimer.
+//
+// 2.  Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation and/or
+// other materials provided with the distribution.
+//
+// 3.  Neither the name of the copyright holder(s) nor the names of any contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission. No license is granted to the trademarks of
+// the copyright holders even if such marks are included in this software.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+import UIKit
+import AVKit
+import AVFoundation
+
+/// `RSDVideoViewController` is a simple view controller for showing a video. The base-class implementation
+/// supports loading a video from a URL, video string, or `RSDResourceTransformer`. It is assumed that
+/// the property will be set for one of these values.
+open class RSDVideoViewController: AVPlayerViewController {
+    
+    /// Convenience method for instantiating a RSDVideoViewController and presenting it on top of the presenter UIViewController
+    open class func instantiateController(action: RSDVideoViewUIAction) -> RSDVideoViewController? {
+        
+        var url: URL? = nil
+        if action.isOnlineResourceURL() {
+            url = URL(string: action.resourceName)
+        } else {
+            do {
+                url = try action.resourceURL().url
+            } catch let err {
+                assertionFailure("Error decoding video resource \(err)")
+                return nil
+            }
+        }
+        
+        guard let urlUnwrapped = url else {
+            assertionFailure("Error decoding video resource, URL cannot be nil")
+            return nil
+        }
+        
+        let player = AVPlayer(url: urlUnwrapped)
+        let playerController = RSDVideoViewController()
+        playerController.player = player
+        return playerController
+    }
+    
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+        self.player?.play()
+    }
+}
```

Comment:
I would suggest extending the resource transformer instead so that this same code can be used elsewhere. :)

```
extension RSDResourceTransformer {
    
    public func fullURL() throws -> URL {
        if self.isOnlineResourceURL() {
            guard let url = URL(string: resourceName) else {
                throw RSDValidationError.unexpectedNullObject("Failed to convert \(resourceName) to a URL.")
            }
            return url
        }
        else {
            return try resourceURL().url
        }
    }
}
```

and then your instantiate method is like this:

```
	    public init(action: RSDVideoViewUIAction)  throws {
                super.init()
	        let url: URL = try action.fullURL()
	        self.player = AVPlayer(url: url)
	    }
    
    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
```

This allows the try/catch to be passed through to the view controller in case a developer ever wants to handle URL loading failures.

Alternatively, do you even need a subclass?  Why not just do this in the step view controller?

```
            // For a videoView action, present an AVPlayerViewController modally.
           do {
               let url = try action.fullURL()
	       let vc = AVPlayerViewController()
               vc.player = AVPlayer(url: url)
               self.present(vc, animated: true) {
                    vc.player?.play()
               }
           }
           catch let err {
                assertionFailure("Failed to get a url for this video. \(err)")
           }
```

The only reason for the weirdness of the web view controller is b/c it is presented inside a navigation controller to allow for the web view controller to be presented modally but still have the web view allow navigation within it. Really, for most of what we do it's a complete overkill but there's always that *one* design that needs navigation. ;)

Also, please confirm that the video should start automatically. Often, especially if there is sound, that is really jarring. I know it's a pet peeve of mine and a reason I will delete an app. Thanks!